File: c++/155. Min Stack.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    MinStack(){
        
    }
    
    void push(int x) {
        st.push(x);
        if (!min_st.empty()) {
            int t = min_st.top();
            if (x <= t)
                min_st.push(x);
        } else
            min_st.push(x);
        
    }
    
    void pop() {
        if (st.empty()) return;
        int t = min_st.top();
        if (t == st.top())
            min_st.pop();
        st.pop();
    }
    
    int top() {
        return st.top();
    }
    
    int getMin() {
        return min_st.top();
    }

private:
    stack<int> st;
    stack<int> min_st;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: python/155. Min Stack.py
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.st = []
        self.min_st = []


    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        self.st.append(x)
        if not self.min_st or (self.min_st and self.min_st[-1] >= x):
            self.min_st.append(x)


    def pop(self):
        """
        :rtype: None
        """
        if not self.st:
            return
            
        x = self.st.pop()
        if x==self.min_st[-1]:
            self.min_st.pop()


    def top(self):
        """
        :rtype: int
        """
        return self.st[-1]


    def getMin(self):
        """
        :rtype: int
        """
        return self.min_st[-1]



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

if __name__ == '__main__':
    st = MinStack()
    st.push(-2)
    st.push(0)
    st.push(-3)
    st.getMin()
    print st.min_st
    st.pop()
    st.top()
    print st.min_st
    st.getMin()






